cmd/restic/cmd_check.go

```diff
@@ -293,11 +287,45 @@ func runCheck(opts CheckOptions, gopts GlobalOptions, args []string) error {
 	}
 
 	switch {
-	case opts.ReadData:
-		doReadData(1, 1)
+	case opts.ReadData, opts.ReadDataSample >= chkr.CountPacks():
+		Verbosef("read all data\n")
+		doReadData(chkr.GetPacks())
 	case opts.ReadDataSubset != "":
 		dataSubset, _ := stringToIntSlice(opts.ReadDataSubset)
-		doReadData(dataSubset[0], dataSubset[1])
+		bucket := dataSubset[0]
+		totalBuckets := dataSubset[1]
+		packs := restic.IDSet{}
+		for pack := range chkr.GetPacks() {
+			// If we ever check more than the first byte
+			// of pack, update totalBucketsMax.
+			if (uint(pack[0]) % totalBuckets) == (bucket - 1) {
+				packs.Insert(pack)
+			}
+		}
+		Verbosef(fmt.Sprintf("read group #%d of %d data packs (out of total %d packs in %d groups)\n", bucket, len(packs), chkr.CountPacks(), totalBuckets))
+		doReadData(packs)
+	case opts.ReadDataSample > 0:
+		rand.Seed(time.Now().UnixNano())
+		maxSkip := int(chkr.CountPacks() - opts.ReadDataSample)
```

Comment:
This algorithm is not correct. Say CountPacks() = 100 and ReadDataSample = 10, then maxSkip = 90. If the RNG returns >=29 three times and >=9 once, the actual sample size is at most 3. This happens with probability (62/91)³ × 82/91 > 28%.

To do this right, you could use [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling).

Reply:
It will always pick a sample of the given size. The number of packs to skip is recomputed after each "skipping".

But you are right, the algorithm is not correct in the sense that it collects a true random sample - in fact packs that are "later" in the `range` will be picked with a much higher probability. As `range` for maps is already kind of random, I ignored this incorrectness :wink: 

Do you now a go package that provides this sampling function? I was looking for one and did not found any - this is why I came with this "hacky" solution...

Reply:
Oh right, I missed the maxSkip-- bit.

Looks like [go-metrics has Vitter's algorithm for reservoir sampling](https://godoc.org/github.com/rcrowley/go-metrics#UniformSample).

Reply:
I have a bunch of random number utilities that I should publish somewhere. I'll see if I can add a reservoir sampler.

Reply:
[Done](https://godoc.org/github.com/greatroar/randstat/sampling). This package could do with a streaming API, but I think the function I have know can be plugged into the current code without too much of a hassle.